peps/pep-9995.rst

```diff
@@ -0,0 +1,218 @@
+PEP: 9995
+Title: Inlined typed dictionaries
+Author: Victorien Plot <contact@vctrn.dev>
+Discussions-To:
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Created: 23-Oct-2024
+Post-History:
+Python-Version: 3.14
+Resolution:
+
+.. highlight:: python
+
+
+Abstract
+========
+
+:pep:`589` defines a `class-based <https://typing.readthedocs.io/en/latest/spec/typeddict.html#class-based-syntax>`_ and a
+:ref:`functional syntax <typing:typeddict-functional-syntax>` to create typed
+dictionaries. In both scenarios, it requires defining a class or assigning to
+a value. In some situations, this can add unnecessary boilerplate, especially
+if the typed dictionary is only used once.
+
+This PEP proposes the addition of a new inlined syntax, by subscripting the
+:class:`~typing.TypedDict` type::
+
+    from typing import TypedDict
+
+    def get_movie() -> TypedDict[{'name': str, 'year': int}]:
+        return {
+            'name': 'Blade Runner',
+            'year': 1982,
+        }
+
+Motivation
+==========
+
+Python dictionaries are an essential data structure of the language. Many
+times, it is used to return or accept structured data in functions. However,
+it can get tedious to define :class:`~typing.TypedDict` classes:
+
+* A typed dictionary requires a name, which might not be relevant.
+* Nested dictionaries requires more than one class definition.
+
+Taking a simple function returning some nested structured data as an example::
+
+    from typing import TypedDict
+
+    class ProductionCompany(TypedDict):
+        name: str
+        location: str
+
+    class Movie(TypedDict):
+        name: str
+        year: int
+        production: ProductionCompany
+
+
+    def get_movie() -> Movie:
+        return {
+            'name': 'Blade Runner',
+            'year': 1982,
+            'production': {
+                'name': 'Warner Bros.',
+                'location': 'California',
+            }
+        }
+
+
+Rationale
+=========
+
+The new inlined syntax can be used to resolve these problems::
+
+    def get_movie() -> TypedDict[{'name': str, year: int, 'production': {'name': str, 'location': str}}]:
+        ...
+
+It is recommended to *only* make use of inlined typed dictionaries when the
+structured data isn't too large, as this can quickly get hard to read.
+
+While less useful (as the functional or even the class-based syntax can be
```

Comment:
I presume that type variables are allowed in an inlined TypedDict. May be worth stating this and/or using a TypeVar in one of the examples just to remind readers (and implementers of type checkers and conformance tests) that it's meant to be supported.

Reply:
I can see the value of supporting something like this:

```python
def func[T](arg: T) -> TypedDict[{'a': T}]: ...


class A[T]:
    attr: TypedDict[{'a': T}]
```

However, does this mean that inlined TypedDicts are implicitly generic when they make use of type variables? (probably not a good idea, as there's no reliable way to detect the use of type variables at runtime so that `Generic` can be added as a base).

Should we only allow usage of type variables if they are not "free" (i.e. already bound to a class or function)?

Reply:
I agree that an inline TypedDict annotation should not constitute its own type variable scope. Any type variable used within an inline TypedDict must be bound to some outer scope (a class, function or type alias).

Reply:
Updated.